File: CS/ChartsDemo.Wpf/ViewModels/ChartDataSourceViewModel.cs
using DevExpress.Mvvm.POCO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows;

namespace ChartsDemo {
    public class ChartDataSourceViewModel<T> {
        public List<ChartDataSource<T>> Sources { get; set; }
        public virtual ChartDataSource<T> SelectedSource { get; set; }
        public virtual IChartAnimationService ChartAnimationService { get { return null; } }

        public void Animate() {
            if(ChartAnimationService != null)
                ChartAnimationService.Animate();
        }
        protected void OnSelectedSourceChanged() {
            Animate();
        }
    }
    public class ChartDataSource<T> {
        public string Description { get; set; }
        public List<T> DataSource { get; set; }
    }
    public static class ChartViewModelFactory {
        #region Scatter Line ViewModels

        public static ChartDataSourceViewModel<Point> CreatePolarScatterLineViewModel() {
            return CreateScatterViewModel(new DegreeScatterFunctionCalculator());
        }
        public static ChartDataSourceViewModel<Point> CreateRadarScatterLineViewModel() {
            return CreateScatterViewModel(new RadianScatterFunctionCalculator());
        }

        static ChartDataSourceViewModel<Point> CreateScatterViewModel(ScatterFunctionCalculatorBase calculator) {
            var viewModel = ViewModelSource.Create<ChartDataSourceViewModel<Point>>();
            viewModel.Sources = new List<ChartDataSource<Point>> {
                new ChartDataSource<Point> { Description = "Archimedean Spiral", DataSource = calculator.CreateArchimedeanSpiralData() },
                new ChartDataSource<Point> { Description = "Polar Rose", DataSource = calculator.CreateRoseData() },
                new ChartDataSource<Point> { Description = "Polar Folium", DataSource = calculator.CreateFoliumData() }
            };
            viewModel.SelectedSource = viewModel.Sources.First();
            return viewModel;
        }

        #endregion
        #region Polar ViewModel

        public static ChartDataSourceViewModel<RangeDataPoint> CreatePolarViewModel() {
            var viewModel = ViewModelSource.Create<ChartDataSourceViewModel<RangeDataPoint>>();
            viewModel.Sources = new List<ChartDataSource<RangeDataPoint>> {
                new ChartDataSource<RangeDataPoint> { Description = "Taubin's Heart", DataSource = PolarFunctionsPointGenerator.CreateTaubinsHeartData() },
                new ChartDataSource<RangeDataPoint> { Description = "Cardioid", DataSource = PolarFunctionsPointGenerator.CreateCardioidData() },
                new ChartDataSource<RangeDataPoint> { Description = "Lemniscate", DataSource = PolarFunctionsPointGenerator.CreateLemniskateData() }
            };
            viewModel.SelectedSource = viewModel.Sources.Last();
            return viewModel;
        }

        #endregion
        #region Cartesian ViewModel

        public static ChartDataSourceViewModel<Point> CreateCartesianViewModel() {
            var viewModel = ViewModelSource.Create<ChartDataSourceViewModel<Point>>();
            viewModel.Sources = new List<ChartDataSource<Point>> {
                new ChartDataSource<Point> { Description = "Archimedean Spiral", DataSource = CartesianFunctionsPointGenerator.CreateArchimedianSpiralPoints() },
                new ChartDataSource<Point> { Description = "Cardioid", DataSource = CartesianFunctionsPointGenerator.CreateCardioidPoints() },
                new ChartDataSource<Point> { Description = "Cartesian Folium", DataSource = CartesianFunctionsPointGenerator.CreateCartesianFoliumPoints() }
            };
            viewModel.SelectedSource = viewModel.Sources.First();
            return viewModel;
        }

        #endregion
    }
}

File: VB/ChartsDemo.Wpf/ViewModels/ChartDataSourceViewModel.vb
Imports DevExpress.Mvvm.POCO
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows

Namespace ChartsDemo
    Public Class ChartDataSourceViewModel(Of T)
        Public Property Sources() As List(Of ChartDataSource(Of T))
        Public Overridable Property SelectedSource() As ChartDataSource(Of T)
        Public Overridable ReadOnly Property ChartAnimationService() As IChartAnimationService
            Get
                Return Nothing
            End Get
        End Property

        Public Sub Animate()
            If ChartAnimationService IsNot Nothing Then
                ChartAnimationService.Animate()
            End If
        End Sub
        Protected Sub OnSelectedSourceChanged()
            Animate()
        End Sub
    End Class
    Public Class ChartDataSource(Of T)
        Public Property Description() As String
        Public Property DataSource() As List(Of T)
    End Class
    Public NotInheritable Class ChartViewModelFactory

        Private Sub New()
        End Sub
        #Region "Scatter Line ViewModels"

        Public Shared Function CreatePolarScatterLineViewModel() As ChartDataSourceViewModel(Of Point)
            Return CreateScatterViewModel(New DegreeScatterFunctionCalculator())
        End Function
        Public Shared Function CreateRadarScatterLineViewModel() As ChartDataSourceViewModel(Of Point)
            Return CreateScatterViewModel(New RadianScatterFunctionCalculator())
        End Function

        Private Shared Function CreateScatterViewModel(ByVal calculator As ScatterFunctionCalculatorBase) As ChartDataSourceViewModel(Of Point)
            Dim viewModel = ViewModelSource.Create(Of ChartDataSourceViewModel(Of Point))()
            viewModel.Sources = New List(Of ChartDataSource(Of Point)) From {
                New ChartDataSource(Of Point) With {.Description = "Archimedean Spiral", .DataSource = calculator.CreateArchimedeanSpiralData()},
                New ChartDataSource(Of Point) With {.Description = "Polar Rose", .DataSource = calculator.CreateRoseData()},
                New ChartDataSource(Of Point) With {.Description = "Polar Folium", .DataSource = calculator.CreateFoliumData()}
            }
            viewModel.SelectedSource = viewModel.Sources.First()
            Return viewModel
        End Function

        #End Region
        #Region "Polar ViewModel"

        Public Shared Function CreatePolarViewModel() As ChartDataSourceViewModel(Of RangeDataPoint)
            Dim viewModel = ViewModelSource.Create(Of ChartDataSourceViewModel(Of RangeDataPoint))()
            viewModel.Sources = New List(Of ChartDataSource(Of RangeDataPoint)) From {
                New ChartDataSource(Of RangeDataPoint) With {.Description = "Taubin's Heart", .DataSource = PolarFunctionsPointGenerator.CreateTaubinsHeartData()},
                New ChartDataSource(Of RangeDataPoint) With {.Description = "Cardioid", .DataSource = PolarFunctionsPointGenerator.CreateCardioidData()},
                New ChartDataSource(Of RangeDataPoint) With {.Description = "Lemniscate", .DataSource = PolarFunctionsPointGenerator.CreateLemniskateData()}
            }
            viewModel.SelectedSource = viewModel.Sources.Last()
            Return viewModel
        End Function

        #End Region
        #Region "Cartesian ViewModel"

        Public Shared Function CreateCartesianViewModel() As ChartDataSourceViewModel(Of Point)
            Dim viewModel = ViewModelSource.Create(Of ChartDataSourceViewModel(Of Point))()
            viewModel.Sources = New List(Of ChartDataSource(Of Point)) From {
                New ChartDataSource(Of Point) With {.Description = "Archimedean Spiral", .DataSource = CartesianFunctionsPointGenerator.CreateArchimedianSpiralPoints()},
                New ChartDataSource(Of Point) With {.Description = "Cardioid", .DataSource = CartesianFunctionsPointGenerator.CreateCardioidPoints()},
                New ChartDataSource(Of Point) With {.Description = "Cartesian Folium", .DataSource = CartesianFunctionsPointGenerator.CreateCartesianFoliumPoints()}
            }
            viewModel.SelectedSource = viewModel.Sources.First()
            Return viewModel
        End Function

        #End Region
    End Class
End Namespace
